### PR TITLE
Rf parallel

### PR DESCRIPTION
**Pull request by yael1994**

### CODE REVIEW

**Comment by yael1994** [comment]
https://trello.com/c/xWGS9AxJ/11-parallel-random-forests

**Comment by yael1994** [comment]
https://trello.com/c/dVXOJFRg/13-stop-random-forest-on-a-value-threshold

**Comment by yael1994** [comment]
https://trello.com/c/eMb9vmYj/16-stop-phase-3-before-random-forest

File: model_fitting/random_forest.py [hunk 1]
```diff
@@ -9,11 +10,18 @@
...
 logger = logging.getLogger('main')
+if os.path.exists('/groups/pupko/orenavr2/'):
+    src_dir = '/groups/pupko/orenavr2/igomeProfilingPipeline/src'
+elif os.path.exists('/Users/Oren/Dropbox/Projects/'):
+    src_dir = '/Users/Oren/Dropbox/Projects/gershoni/src'
+else:
+    src_dir = '.'
+sys.path.insert(0, src_dir)
 
+from auxiliaries.pipeline_auxiliaries import *
```
Comment (shacharmo) [inline model_fitting/random_forest.py]: Don't use implicit imports

File: model_fitting/random_forest.py [hunk 2]
```diff
@@ -90,35 +98,21 @@ def sample_configurations(hyperparameters_grid, num_of_configurations_to_sample,
...
     # transform the data for better contrast in the visualization
-    if hits_data:  # hits data
-        train_data = np.log2(df+1)  # pseudo counts
-    elif use_tfidf: # tfidf data
-        train_data = -np.log2(df+0.0001)  # avoid 0
-    else:  # p-values data
-        train_data = -np.log2(df)
-
-    cm = sns.clustermap(df, cmap="Blues", col_cluster=False, yticklabels=True)
+    train_data = np.log2(df+1) if hits_data else df
```
Comment (shacharmo) [inline model_fitting/random_forest.py]: Is this correct? What happens to p-values?

File: model_fitting/random_forest.py [hunk 3]
```diff
@@ -90,35 +98,21 @@ def sample_configurations(hyperparameters_grid, num_of_configurations_to_sample,
...
-        train_data = -np.log2(df)
-
-    cm = sns.clustermap(df, cmap="Blues", col_cluster=False, yticklabels=True)
+    train_data = np.log2(df+1) if hits_data else df 
+    
+    
+    #elif use_tfidf: # tfidf data
+    #    train_data = -np.log2(df+0.0001) 
+    #else:  # p-values data
+    #    train_data = -np.log2(df)
```
Comment (shacharmo) [inline model_fitting/random_forest.py]: Remove unused code

**Review by shacharmo — CHANGES_REQUESTED** [review]
Was this thoroughly tested? are we getting same or better results then before quicker?
Are the results reproducible, i.e. running the same experiments leads to same results (hint: check for seed in train_random_forest)

File: IgOmeProfiling_pipeline.py [hunk 4]
```diff
@@ -161,14 +166,18 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
...
+    parser.add_argument('--min_value_error_random_forest', default=0, type=float, help='A min value for error that the run can stop in the random forest')
     parser.add_argument('--rank_method', choices=['pval', 'tfidf', 'shuffles'], default='pval', help='Motifs ranking method')
     parser.add_argument('--tfidf_method', choices=['boolean', 'terms', 'log', 'augmented'], default='boolean', help='TF-IDF method')
     parser.add_argument('--tfidf_factor', type=float, default=0.5, help='TF-IDF augmented method factor (0-1)')
     parser.add_argument('--shuffles', default=5, type=int, help='Number of controlled shuffles permutations')
     parser.add_argument('--shuffles_percent', default=0.2, type=float, help='Percent from shuffle with greatest number of hits (0-1)')
     parser.add_argument('--shuffles_digits', default=2, type=int, help='Number of digits after the point to print in scanning files.')
-
+    parser.add_argument('--seed_random_forest_classifier', defual=123 , type=int, help='A number for create the random forest stable when run the same configuration')
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: defaul => default
Change to `random_forest_seed`.
Should be: Random seed value for generating random forest configurations

File: IgOmeProfiling_pipeline.py [hunk 5]
```diff
@@ -161,14 +166,18 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
     parser.add_argument('-m', '--mapitope', action='store_true', help='use mapitope encoding')
 
     # optional parameters for the modelling step
+    parser.add_argument('--stop_random_forest', action='store_true', help='A boolean flag for mark if we need to run the random forest')
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: change to `stop_before_random_forest` or `stop_after_motifs_ranking`

File: IgOmeProfiling_pipeline.py [hunk 6]
```diff
@@ -161,14 +166,18 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
...
     parser.add_argument('--number_of_random_pssms', default=100, type=int, help='Number of pssm permutations')
+    parser.add_argument('--number_parallel_random_forest', default=20, type=int, help='How many random forest to run in parallel')
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: Should be: How many rando forest **configurations** to run in parallel

File: IgOmeProfiling_pipeline.py [hunk 7]
```diff
@@ -161,14 +166,18 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
...
+    parser.add_argument('--min_value_error_random_forest', default=0, type=float, help='A min value for error that the run can stop in the random forest')
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: Should be: A random forest model error value for convergence allowing to stop early

File: IgOmeProfiling_pipeline.py [hunk 8]
```diff
@@ -195,6 +204,7 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
                  args.left_construct, args.right_construct, args.max_mismatches_allowed, args.min_sequencing_quality, args.minimal_length_required, True if args.gz else False,
                  args.max_msas_per_sample, args.max_msas_per_bc,
                  args.max_number_of_cluster_members_per_sample, args.max_number_of_cluster_members_per_bc,
-                 args.allowed_gap_frequency, concurrent_cutoffs, args.meme_split_size, args.mapitope, args.number_of_random_pssms,
+                 args.allowed_gap_frequency, concurrent_cutoffs, args.meme_split_size, args.mapitope,True if args.stop_random_forest else False,
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: Some arguments are boolean, no need for `True if arg else False`.
Fix it for every boolean argument in entire code

File: IgOmeProfiling_pipeline.py [hunk 9]
```diff
@@ -195,6 +204,7 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
...
+                 args.number_of_random_pssms, args.number_parallel_random_forest, args.min_value_error_random_forest,
                  args.rank_method, args.tfidf_method, args.tfidf_factor, args.shuffles, args.shuffles_percent, args.shuffles_digits,
-                 run_summary_path, error_path, args.queue, True if args.verbose else False, sys.argv)
+                 args.seed_random_forest_classifier, run_summary_path, error_path, args.queue, True if args.verbose else False, sys.argv)
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: Always end file with a blank line (best practice)

File: model_fitting/random_forest.py (hunk 2 above)
Comment (shacharmo) [reply model_fitting/random_forest.py]: You resolved the conversion but didn't reply.
This doesn't support p-value which has different normalization (-np.log2(df + epsilon))

File: model_fitting/train_random_forest.py [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+def generate_roc_curve(X, y, classifier, number_of_features, output_path):
+    ax = plt.gca()
+    plot_roc_curve(classifier, X, y, **{'marker': 'o'}, ax=ax)
+    plt.savefig(f"{output_path}/roc_curve_{number_of_features}.png", format='png', bbox_inches="tight")
+    plt.close()
+
+
+def generate_heat_map(df, number_of_features, hits_data, number_of_samples, output_path):
+    train_data = np.log2(df+1) if hits_data else df
```
Comment (shacharmo) [inline model_fitting/train_random_forest.py:50]: Same comment as above, this doesn't support p-value correctly

File: model_fitting/train_random_forest.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('configuration_path', type=str, help='A dictionary of the configuration to this model ')
+    parser.add_argument('csv_file_path', type=str, help='A csv file with data matrix to model')
+    parser.add_argument('is_hits_data', type=bool, help='if is hits data or pvalue data')
+    parser.add_argument('output_path_i', type=str, help='A path for the results of this model')
+    parser.add_argument('model_number', type=str, help='number of model')
+    parser.add_argument('done_file_path', type=str, help='A path to a file that signals that the script finished running successfully')
+    parser.add_argument('--seed_random_forest_classifier', defual=123 , type=int, help='A number for create the random forest stable when run the same configuration')
+    #parser.add_argument('--tfidf', action='store_true', help="Are inputs from TF-IDF (avoid log(0))")
```
Comment (shacharmo) [inline model_fitting/train_random_forest.py]: Remove unused code

File: model_fitting/train_random_forest.py [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+    args = parser.parse_args()
+    import logging
+    if args.verbose:
+        logging.basicConfig(level=logging.DEBUG)
+    else:
+        logging.basicConfig(level=logging.INFO)
+    logger = logging.getLogger('main')
+
+   
+    pre_train(args.configuration_path, args.csv_file_path, args.is_hits_data, args.output_path_i, args.model_number, args.done_file_path, args.seed_random_forest_classifier, args.cv_num_of_splits, argv=sys.argv)
```
Comment (shacharmo) [inline model_fitting/train_random_forest.py]: Add blank line at end of file (best practice)

**Review by shacharmo — CHANGES_REQUESTED** [review]
Also, fix conflicts

File: model_fitting/train_random_forest.py (hunk 10 above)
Comment (yael1994) [reply model_fitting/train_random_forest.py:50]: I thought it was decided after we showed to the client the logs to leave only for hits.

Comment (shacharmo) [reply model_fitting/train_random_forest.py:50]: Even if this is true (and I'm not sure they wanted it for p-val), the p-val has opposite values, i.e. 0 is "best" value

File: IgOmeProfiling_pipeline.py [hunk 13]
```diff
@@ -166,17 +166,17 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
     parser.add_argument('-m', '--mapitope', action='store_true', help='use mapitope encoding')
 
     # optional parameters for the modelling step
-    parser.add_argument('--stop_random_forest', action='store_true', help='A boolean flag for mark if we need to run the random forest')
+    parser.add_argument('--stop_before_random_forest', action='store_true', help='A boolean flag for mark if we need to run the random forest')
     parser.add_argument('--number_of_random_pssms', default=100, type=int, help='Number of pssm permutations')
-    parser.add_argument('--number_parallel_random_forest', default=20, type=int, help='How many random forest to run in parallel')
-    parser.add_argument('--min_value_error_random_forest', default=0, type=float, help='A min value for error that the run can stop in the random forest')
+    parser.add_argument('--number_parallel_random_forest', default=20, type=int, help='How many rando forest configurations to run in parallel')
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: rando => random
I know this is a typo in the comment but please read and validate 😃

File: model_fitting/random_forest.py [hunk 14]
```diff
@@ -299,12 +295,12 @@ def save_configuration_to_txt_file(sampled_configuration, output_path_i):
     parser.add_argument('logs_dir', help='A path for the log dir')
     parser.add_argument('error_path', help='Path for error file')
     parser.add_argument('--num_of_configurations_to_sample', default=100, type=int, help='How many random configurations of hyperparameters should be sampled?')
-    parser.add_argument('--number_parallel_random_forest', default=20, type=int, help='How many random forest to run in parallel')
-    parser.add_argument('--min_value_error_random_forest', default=0, type=float, help='A min value for error that the run can stop')
+    parser.add_argument('--number_parallel_random_forest', default=20, type=int, help='How many rando forest configurations to run in parallel')
```
Comment (shacharmo) [inline model_fitting/random_forest.py]: Same comment, rando => random

**Review by shacharmo — CHANGES_REQUESTED** [review]
Still need to properly address p-val.
Note that values are opposite, i.e., value of 0 for number of hits greater than all shuffles.

Regarding log scale, maybe add it as controllable parameter.
If so, do it in another PR

**Review by shacharmo — APPROVED** [review]

**Pull request merged**